### PR TITLE
[SPARK-6299][CORE] ClassNotFoundException in standalone mode when running groupByKey with class defined in REPL

### PR DESCRIPTION
**Pull request by swkimme**
```
case class ClassA(value: String)
val rdd = sc.parallelize(List(("k1", ClassA("v1")), ("k1", ClassA("v2")) ))
rdd.groupByKey.collect
```

This code used to be throw exception in spark-shell, because while shuffling `JavaSerializer`uses `defaultClassLoader` which was defined like `env.serializer.setDefaultClassLoader(urlClassLoader)`.

It should be `env.serializer.setDefaultClassLoader(replClassLoader)`, like 

```
    override def run() {
      val deserializeStartTime = System.currentTimeMillis()
      Thread.currentThread.setContextClassLoader(replClassLoader)
```

in TaskRunner. 

When `replClassLoader` cannot be defined, it's identical with `urlClassLoader`

### CODE REVIEW

**Comment by srowen** [comment]
ok to test

**Comment by shaneknapp** [comment]
this was me, i'll retrigger the job once maintenance is over.

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by rxin** [comment]
LGTM pending tests.

**Comment by rxin** [comment]
Actually would be great to add a test case to the repl suite. Can you try constructing a test case?

**Comment by swkimme** [comment]
@rxin 
Sure, I'll try to add some test.

**Comment by swkimme** [comment]
@rxin 
I tried to add a simple test like 
```
  test("collecting objects of class defined in repl - shuffling") {
    val output = runInterpreter("local-cluster[1,1,512]",
      """
        |case class Foo(i: Int)
        |val list = List((1, Foo(1)), (1, Foo(2)))
        |val ret = sc.parallelize(list).groupByKey.collect
      """.stripMargin)
    assertDoesNotContain("error:", output)
    assertDoesNotContain("Exception", output)
    assertContains("ret: Array[(Int, Iterable[Foo])] = Array((1,", output)
  }
```
But it failed. (Works well in real cluster)

Am I doing something wrong?

**Comment by rxin** [comment]
You need to do sbt/sbt assembly/assembly before running that tests, since the repl forks a process to launch Spark.

File: repl/scala-2.10/src/test/scala/org/apache/spark/repl/ReplSuite.scala [hunk 1]
```diff
@@ -315,4 +315,16 @@ class ReplSuite extends FunSuite {
     assertDoesNotContain("Exception", output)
     assertContains("ret: Array[Foo] = Array(Foo(1),", output)
   }
+  
+  test("collecting objects of class defined in repl - shuffling") {
+    val output = runInterpreter("local-cluster[1,1,512]",
+      """
+        |case class Foo(i: Int)
+        |val list = List((1, Foo(1)), (1, Foo(2)))
+        |val ret = sc.parallelize(list).groupByKey.collect
```
Comment (rxin) [inline repl/scala-2.10/src/test/scala/org/apache/spark/repl/ReplSuite.scala]: can u add parentheses to groupByKey and collect, i.e. groupByKey().collect()

File: repl/scala-2.11/src/test/scala/org/apache/spark/repl/ReplSuite.scala [hunk 2]
```diff
@@ -323,4 +323,16 @@ class ReplSuite extends FunSuite {
     assertDoesNotContain("Exception", output)
     assertContains("ret: Array[Foo] = Array(Foo(1),", output)
   }
+  
+  test("collecting objects of class defined in repl - shuffling") {
+    val output = runInterpreter("local-cluster[1,1,512]",
+      """
+        |case class Foo(i: Int)
+        |val list = List((1, Foo(1)), (1, Foo(2)))
+        |val ret = sc.parallelize(list).groupByKey.collect
```
Comment (rxin) [inline repl/scala-2.11/src/test/scala/org/apache/spark/repl/ReplSuite.scala]: here too

**Comment by swkimme** [comment]
@rxin 
That would be easy, btw, other test are written like ```collect```, not ```collect()```
Do I change other lines?

**Comment by rxin** [comment]
Would be great to change them (just this file) while you are at it.  Thanks!

**Comment by rxin** [comment]
Thanks. I've merged this in master & branch-1.3. It will be fixed in 1.3.1.

**Comment by swkimme** [comment]
@rxin 
Great, thanks!

**Pull request closed without merging**